### ISSUE TITLE
PropTypes validation invalid when using function methods

### ISSUE THREAD

**Issue by globexdesigns** [yannickcr/eslint-plugin-react#134]
The following code:

```
<a onClick={this.props.something.bind(this, 1)}
```

Fails against the propTypes required definition, saying that `this.props.something.bind` isn't defined. But `this.props.something` is already in the propTypes.

This started happening as of 2.6.0

**Comment by yannickcr** [comment]
Is `this.props.something` correctly defined as a function in `propTypes` ?

I have no problem with the following code.

test.js

```js
var Hello = React.createClass({
  propTypes: {
    something: React.PropTypes.func.isRequired
  },
  render: function() {
    return <a onClick={this.props.something.bind(this, 1)}>Hello</a>;
  }
});
```

.eslintrc

```json
{
  "env": {
    "browser": true,
    "node": true
  },
  "ecmaFeatures": {
    "jsx": true
  },
  "plugins": [
    "react"
  ],
  "rules": {
    "react/prop-types": 2
  }
}
```

**Comment by globexdesigns** [comment]
I can't reproduce anymore either. I'll reopen if I come across again.

**Comment by globexdesigns** [comment]
Found a very similar issue using the following code:

```
import React, {Component} from 'react/addons';

export default class Test extends Component {
	componentWillUpdate (nextProps, nextState) {
		console.log(this.props.children.props.children);
	}

	render () {
		return <span/>;
	}
}
```

This results in the following error:

`'children.props` is missing in props validation for Test at line 5 col 9`

**Comment by yannickcr** [comment]
`this.props.children` require to be defined in `propTypes` too.

For example a fixed component would be:

```js
import React, {Component} from 'react/addons';

export default class Test extends Component {
    static propTypes = {
        children: React.PropTypes.node
    }
    componentWillUpdate (nextProps, nextState) {
        console.log(this.props.children.props.children);
    }

    render () {
        return <span/>;
    }
}
```

You can find more information in [the documentation](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/prop-types.md#as-for-exceptions).

You can also use the [ignore option](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/prop-types.md#rule-options) if you really do not want to declare `children` in propTypes.

**Comment by globexdesigns** [comment]
Actually my `ignore` already has children -- which is why I think this is a bug. My rule is defined as:

```
react/prop-types: [2, {"ignore": ["className", "children", "id", "onClick", "style"]}]
```

It looks like anything that user this syntax `this.props.foo.bar` is being validated for `foo` and `foo.bar` now.

**Comment by yannickcr** [comment]
There was a bug on the `ignore` option (#135), it was fixed in `v2.6.3`. Can you upgrade to the latest version and tell me if it fixes your issue ? Thanks.

**Comment by globexdesigns** [comment]
Unfortunately no. The following code still fails:

```
import React, {Component} from 'react/addons';

export default class Test extends Component {
    static propTypes = {
        test: React.PropTypes.object
    }

    render () {
        console.log(this.props.test.someKey)
        return <span/>;
    }
}
```

Gives me:

```
 9:32  error    'test.someKey' is missing in props validation     react/prop-types
```

**Comment by mathieumg** [comment]
But then, it *is* missing.

**Comment by globexdesigns** [comment]
Are you suggesting that I need to define all possible keys for my prop objects? That's... a bit extreme, no? I understand if the prop type is is "shape", but if I say it's an "object" then the assumption is that it's a varying object type.

**Comment by mathieumg** [comment]
Perhaps, but then again, if you use it in that very component, it's a good assumption to make it should be explicited in the `propTypes`, no?

**Comment by globexdesigns** [comment]
Ok, thinking about this some more - I don't disagree.

However, it's going to require PropTypes definitions to be a lot more precise. Basically `PropType.object` becomes useless because as soon as you use any key from that object -- the PropType validation will fail. So you have to either use `PropType.shape` or `PropType.instanceOf`.

**Comment by mathieumg** [comment]
> PropType.object becomes useless because as soon as you use any key from that object

Fair point. So I imagine, if you use a property which is itself declared as such (`PropTypes.object`), it should still error?
